ktor-server/ktor-server-host-common/jvm/src/io/ktor/server/engine/ApplicationEngineEnvironmentReloading.kt

```diff
@@ -289,7 +290,13 @@ public class ApplicationEngineEnvironmentReloading(
     }
 
     private fun instantiateAndConfigureApplication(currentClassLoader: ClassLoader): Application {
-        val newInstance = Application(this)
+        val newInstance = if (recreateInstance || _applicationInstance == null) {
```

Comment:
What will happen if I call `application.coroutineContext[Job]?.invokeOnCompletion { println("foo") }` before application actually starts? The same question for `intercept` method

Reply:
I've added tests to check these cases.